File: dsa/queues/elixir/lifo_queue.ex
defmodule LIFOQueue do
  def new, do: {[], []}

  # O(1)
  def enqueue({primary, amortized}, element), do: push({primary, amortized}, element)

  # O(n) -> O(1) amortized
  def dequeue({[], []}), do: pop({[], []})
  def dequeue({primary, [head | tail]}), do: pop({primary, [head | tail]})
  def dequeue({primary, []}), do: pop({primary, []})

  # O(1)
  defp push({primary, amortized}, element), do: {[element | primary], amortized}

  # O(n) -> O(1) amortized
  defp pop({[], []}), do: {nil, {[], []}}
  defp pop({primary, [head | tail]}), do: {head, {primary, tail}}
  defp pop({primary, []}), do: pop({[], reverse(primary, [])})

  defp reverse([], acc), do: acc
  defp reverse([head | tail], acc), do: reverse(tail, [head | acc])
end

{primary, amortized} = LIFOQueue.new

{primary, amortized} = LIFOQueue.enqueue({primary, amortized}, 1)
{primary, amortized} = LIFOQueue.enqueue({primary, amortized}, 2)
{primary, amortized} = LIFOQueue.enqueue({primary, amortized}, 3)

{element, {primary, amortized}} = LIFOQueue.dequeue({primary, amortized})
IO.inspect(element)

{element, {primary, amortized}} = LIFOQueue.dequeue({primary, amortized})
IO.inspect(element)

{element, {primary, amortized}} = LIFOQueue.dequeue({primary, amortized})
IO.inspect(element)

{primary, amortized} = LIFOQueue.enqueue({primary, amortized}, 4)

{element, {_, _}} = LIFOQueue.dequeue({primary, amortized})
IO.inspect(element)

File: dsa/queues/rust/lifo_queue.rs
struct LIFOQueue<T> {
    stack: Vec<T>,
    amortized: Vec<T>
}

impl<T> LIFOQueue<T> {
    fn new() -> Self {
        Self { stack: Vec::new(), amortized: Vec::new() }
    }

    fn enqueue(&mut self, item: T) {
        self.stack.push(item)
    }

    fn dequeue(&mut self) -> Option<T> {
        if self.amortized.is_empty() {
            while let Some(item) = self.stack.pop() {
                self.amortized.push(item);
            }
        }
        self.amortized.pop()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lifo_queue() {
        let mut queue = LIFOQueue::new();

        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);

        assert_eq!(queue.dequeue(), Some(1));
        assert_eq!(queue.dequeue(), Some(2));
        assert_eq!(queue.dequeue(), Some(3));
        assert_eq!(queue.dequeue(), None);
    }
}
